### ISSUE TITLE
Question about AdversarialDiscriminatorLoss in released checkpoint and its impact on batch effect correction

### ISSUE THREAD

**Issue by BlenderWang9487** [cantinilab/scPRINT#2]
Hi,

Firstly, thank you for your great work. You have opened up a new single cell transformer pretraining paradigm, which is very inspiring.

I'd like to ask if the checkpoint you released was trained using AdversarialDiscriminatorLoss? 
This loss isn't mentioned in your preprint, but in the state dict of the checkpoint, I can find the weight for this layer: 

`grad_reverse_discriminator_loss.decoder.0.weight ...`

I'm wondering if you've done an ablation study on this loss? 
What's the impact of this loss on batch effect correction?

Thank you.

**Comment by jkobject** [comment]
Hi @BlenderWang9487!

You are pointing out a good question. I have tried multiple ideas to make the model learn to remove the batch effect, and "AdversarialDiscriminatorLoss" was one of them. However it did not seem to work well. 

Thus, I did not use it in my pre-training. You might still see the weights because I think I am initializing them even if they are not used (something I should remove from the init(), to be honest). But they should contain random initialization values.

You can see more about the pertaining in my wandb logs here: https://wandb.ai/ml4ig/scprint_scale/reports/scPRINT-trainings--Vmlldzo4ODIxMjgx?accessToken=80metwx7b08hhourotpskdyaxiflq700xzmzymr6scvkp69agybt79l341tv68hp (also listed in the readme)

One thing to note regarding removing batch effects is that it was not the main aim of the paper; thus, I did not spend too much time trying to solve it. I now have a couple more ideas to try to make it work, and some will be quick to implement and test. I will start to work on this in a week.  But I would be happy to discuss them more with you if you want to collaborate on this.

**Comment by BlenderWang9487** [comment]
Thank you for your response.
It's fascinating to hear about your experimentation process.

I'm interested in this topic because my project also involves working on a multi-species single cell transformer, which is also utilize the ESM protein model's embedding like UCE. My focus is on reducing the differences between similar cell types from different species in the embedding space. The goal is to train a model on certain species that can be directly tested on new, unseen species.

Your approach to batch effect correction caught my attention while I was reviewing your codebase. I was considering applying a similar method to species labels if it proved effective for batch effect correction.

Although you mentioned it doesn't seem to work well currently, I wonder if this might be due to the nature of scRNA-seq experiments. Typically, these datasets focus on a single tissue or organ, so the differences between datasets might primarily reflect biological variations rather than batch effects. This could potentially make it easier for the Discriminator to distinguish between biologically diverse datasets while overlooking the subtle batch effects within similar cell types (e.g., lung cells from batch A vs. batch B). Could this be hindering the effectiveness of the reversed gradient in learning batch effect correction?

Despite the current challenges, I find your idea of using AdversarialDiscriminatorLoss quite cool. 
Thank you again for your openness in sharing your process and ideas!

**Comment by jkobject** [comment]
This is a cool project! Let me know if you want to work starting from the scPRINT codebase, I would be happy to help and provide any guidelines. Adding an adversarial species classifier to the cell type label is easy.

Also, I would be interested in seeing what happens if you use scPRINT in the decoding scheme, setting a cell type embedding from a mouse species but with the embedding of human species and using human genes!

**Comment by jkobject** [comment]
marking this as closed now.
